src/services/refactors/extractMethod.ts

```diff
@@ -0,0 +1,1154 @@
+                return { body: createBlock(body.statements, /*multLine*/ true), returnValueProperty: undefined };
+            }
+            let returnValueProperty: string;
+            const statements = createNodeArray(isBlock(body) ? body.statements.slice(0) : [isStatement(body) ? body : createReturn(<Expression>body)]);
+            // rewrite body if either there are writes that should be propagated back via return statements or there are substitutions
+            if (writes || substitutions.size) {
+                const rewrittenStatements = visitNodes(statements, visitor).slice();
+                if (writes && !(range.facts & RangeFacts.HasReturn) && isStatement(body)) {
+                    // add return at the end to propagate writes back in case if control flow falls out of the function body
+                    // it is ok to know that range has at least one return since it we only allow unconditional returns
+                    const assignments = getPropertyAssignmentsForWrites(writes);
+                    if (assignments.length === 1) {
+                        rewrittenStatements.push(createReturn(assignments[0].name));
+                    }
+                    else {
+                        rewrittenStatements.push(createReturn(createObjectLiteral(assignments)));
+                    }
+                }
+                return { body: createBlock(rewrittenStatements, /*multiLine*/ true), returnValueProperty };
+            }
+            else {
+                return { body: createBlock(statements, /*multiLine*/ true), returnValueProperty: undefined };
+            }
+
+            function visitor(node: Node): VisitResult<Node> {
+                if (node.kind === SyntaxKind.ReturnStatement && writes) {
+                    const assignments: ObjectLiteralElementLike[] = getPropertyAssignmentsForWrites(writes);
+                    if ((<ReturnStatement>node).expression) {
+                        if (!returnValueProperty) {
+                            returnValueProperty = generateReturnValueProperty();
+                        }
+                        assignments.unshift(createPropertyAssignment(returnValueProperty, visitNode((<ReturnStatement>node).expression, visitor)));
+                    }
+                    if (assignments.length === 1) {
+                        return createReturn(assignments[0].name as Expression);
+                    }
+                    else {
+                        return createReturn(createObjectLiteral(assignments));
+                    }
+                }
+                else {
+                    const substitution = substitutions.get(getNodeId(node).toString());
+                    return substitution || visitEachChild(node, visitor, nullTransformationContext);
+                }
+            }
+        }
+    }
+
+    function isModuleBlock(n: Node): n is ModuleBlock {

```

Comment:
I would have guessed there was a family of such type guards in a shared location.